lib/semantics/resolve-names.cc

```diff
@@ -5915,6 +5893,315 @@ class DeferredCheckVisitor {
+  switch (beginDir.v) {
+  case parser::OmpLoopDirective::Directive::Distribute:
+    PushContext(beginDir.source, OmpDirective::DISTRIBUTE);
+    break;
+  case parser::OmpLoopDirective::Directive::Do:
+    PushContext(beginDir.source, OmpDirective::DO);
+    break;
+  case parser::OmpLoopDirective::Directive::DoSimd:
+    PushContext(beginDir.source, OmpDirective::DO_SIMD);
+    break;
+  case parser::OmpLoopDirective::Directive::ParallelDo:
+    PushContext(beginDir.source, OmpDirective::PARALLEL_DO);
+    break;
+  case parser::OmpLoopDirective::Directive::ParallelDoSimd:
+    PushContext(beginDir.source, OmpDirective::PARALLEL_DO_SIMD);
+    break;
+  case parser::OmpLoopDirective::Directive::Simd:
+    PushContext(beginDir.source, OmpDirective::SIMD);
+    break;
+  case parser::OmpLoopDirective::Directive::Taskloop:
+    PushContext(beginDir.source, OmpDirective::TASKLOOP);
+    break;
+  case parser::OmpLoopDirective::Directive::TaskloopSimd:
+    PushContext(beginDir.source, OmpDirective::TASKLOOP_SIMD);
+    break;
+  default:
+    // TODO others
+    break;
+  }
+  ClearDataSharingAttributeObjects();
+  return true;
+}
+
+bool OmpAttributeVisitor::Pre(const parser::OpenMPSectionsConstruct &x) {
+  const auto &beginSectionsDir{
+      std::get<parser::OmpBeginSectionsDirective>(x.t)};
+  const auto &beginDir{
+      std::get<parser::OmpSectionsDirective>(beginSectionsDir.t)};
+  switch (beginDir.v) {
+  case parser::OmpSectionsDirective::Directive::ParallelSections:
+    PushContext(beginDir.source, OmpDirective::PARALLEL_SECTIONS);
+    break;
+  case parser::OmpSectionsDirective::Directive::Sections:
+    PushContext(beginDir.source, OmpDirective::SECTIONS);
+    break;
+  default: break;
+  }
+  ClearDataSharingAttributeObjects();
+  return true;
+}
+
+bool OmpAttributeVisitor::Pre(const parser::OpenMPThreadprivate &x) {
+  PushContext(x.source, OmpDirective::THREADPRIVATE);
+  const auto &list{std::get<parser::OmpObjectList>(x.t)};
+  ResolveOmpObjectList(list, Symbol::Flag::OmpThreadprivate);
+  return false;
+}
+
+void OmpAttributeVisitor::Post(const parser::OmpDefaultClause &x) {
+  if (!ompContext_.empty()) {
+    switch (x.v) {
+    case parser::OmpDefaultClause::Type::Private:
+      SetContextDefaultDSA(Symbol::Flag::OmpPrivate);
+      break;
+    case parser::OmpDefaultClause::Type::Firstprivate:
+      SetContextDefaultDSA(Symbol::Flag::OmpFirstPrivate);
+      break;
+    case parser::OmpDefaultClause::Type::Shared:
+      SetContextDefaultDSA(Symbol::Flag::OmpShared);
+      break;
+    case parser::OmpDefaultClause::Type::None:
+      SetContextDefaultDSA(Symbol::Flag::OmpNone);
+      break;
+    }
+  }
+}
+
+// For OpenMP constructs, check all the data-refs within the constructs
+// and adjust the symbol for each Name if necessary
+void OmpAttributeVisitor::Post(const parser::Name &name) {
+  if (name.symbol && !ompContext_.empty() && GetContext().withinConstruct) {
+    auto *symbol{name.symbol};
```

Comment:
It would be a little clearer to declare symbol outside the if and test symbol in the if condition.
As it is, it looks like you're dereferencing symbol without checking for null.